snapcraft/plugins/_python/_pip.py

```diff
@@ -0,0 +1,263 @@
+        the packages should have already been downloaded via `self.download`.
+
+        :param list args: Arguments for `pip wheel`. Package names, for
+                          example, or the path to a dir containing a setup.py.
+
+        :return: List of paths to each wheel that was built.
+        :rtype: list
+        """
+        if not args:
+            return  # No wheels to build
+
+        wheels = []
+        with tempfile.TemporaryDirectory() as temp_dir:
+
+            # Using pip with a few special parameters:
+            #
+            # --no-index: Don't hit pypi, assume the packages are already
+            #             downloaded (i.e. by using `self.download()`)
+            # --find-links: Provide the directory into which the packages
+            #               should have already been fetched
+            # --wheel-dir: Build wheels into a temporary working area rather
+            #              rather than cwd. We'll copy them over. FIXME: Why
+            #              not just build them straight in the package_dir?
+            self._run(['wheel', '--no-index', '--find-links',
+                       self._python_package_dir, '--wheel-dir',
+                       temp_dir] + list(args), **kwargs)
+            wheels = os.listdir(temp_dir)
+            for wheel in wheels:
+                file_utils.link_or_copy(
+                    os.path.join(temp_dir, wheel),
+                    os.path.join(self._python_package_dir, wheel))
+
+        return [os.path.join(self._python_package_dir, wheel)
+                for wheel in wheels]
+
+    def list(self):
+        """Determine which packages have been installed.
+
+        :return: Dict of installed python packages and their versions
+        :rtype: dict
+        """
+        output = self._run(['list', '--format=json'])
+        packages = collections.OrderedDict()
+        try:
+            json_output = json.loads(
+                    output, object_pairs_hook=collections.OrderedDict)
+        except json.decoder.JSONDecodeError as e:
+            raise errors.PipListInvalidJsonError(output) from e
+
+        for package in json_output:
+            if 'name' not in package:
+                raise errors.PipListMissingFieldError('name', output)
+            if 'version' not in package:
+                raise errors.PipListMissingFieldError('version', output)
+            packages[package['name']] = package['version']
+        return packages
+
+    def clean_packages(self):
+        """Remove the package cache."""
+        with contextlib.suppress(FileNotFoundError):
+            shutil.rmtree(self._python_package_dir)
+
+    def env(self):
+        """The environment used by pip.
+
+        This function is only useful if you happen to need to call into pip's
+        environment without using the API otherwise made available here (e.g.
+        calling the setup.py directly instead of with pip).
```

Comment:
I got lost here. This sounds like an internal method _env. If somebody needs to call setup.py without pip, why would he use the pip class?

Reply:
Yeah this seems weird when reviewing in isolation, so to explain allow me to point you to what WILL use it. The Python plugin uses pip to download and install packages from `python-packages`, the setup.py, requirements.txt, etc. However, in addition to that, it also runs the setup.py manually, [here](https://github.com/snapcore/snapcraft/blob/master/snapcraft/plugins/python.py#L292). That needs to execute within the same environment, and yet as it explicitly doesn't use pip, this refactor does not extract it from the plugin.

Do you see a cleaner way to do that?

Reply:
Got it.

Reply:
Might want to add something like, e.g. when used in scriptlets.

Reply:
@sergiusens does this help in scriptlets? You still can't just run `pip` and expect it to work. We need some sort of binary wrapper for that, no?